src/Raven.Server/Documents/Indexes/ConcurrentLruRegexCache.cs

```diff
@@ -0,0 +1,126 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.Linq;
+using System.Text;
+using System.Text.RegularExpressions;
+using System.Threading;
+using Sparrow.Collections.LockFree;
+
+namespace Raven.Server.Documents.Indexes
+{
+    public class ConcurrentLruRegexCache
+    {
+        public ConcurrentLruRegexCache(int capacity)
+        {
+            _capacity = capacity;
+            _halfCapacity = _capacity / 2;
+            Debug.Assert(_halfCapacity > 0);
+        }
+        private readonly ConcurrentDictionary<string, ConcurrentLruRegexCacheNode> _regexCache = new ConcurrentDictionary<string, ConcurrentLruRegexCacheNode>();
+        private readonly int _capacity;
+        private long _count;
+        private bool _neverCompile;
+        private DateTime _lastClearTime = DateTime.MinValue;
+        private readonly int _halfCapacity;
+
+        public Regex Get(string pattern)
+        {
+            if (!_regexCache.TryGetValue(pattern, out var result))
+                return GetUnlikely(pattern); // create it
+
+            // we don't care about this until the cache is over half full,
+            // so this should be called if user is using a reasonable number
+            // of regex queries.
+            if (_count > _halfCapacity)
+                Updatetimestamp(result);
+            return result.RegexLazy.Value; 
+        }
+
+        private static void Updatetimestamp(ConcurrentLruRegexCacheNode result)
+        {
+            var timestamp = Stopwatch.GetTimestamp();
+
+            var resultTimestamp = result.Timestamp;
+            // here we want to avoid updating this using Interlocked on a highly
+            // used query, so we limit it to once per 10,000 ticks, which should
+            // still be good enough if we need to start evicting
+            while (resultTimestamp - timestamp > 10_000)
```

Comment:
Actually need to update the timestamp as well.

Reply:
i'm not following which time stamp you refer to here

Reply:
yes